lua/bookmarks/tree/context.lua

```diff
@@ -0,0 +1,90 @@
+local domain = require("bookmarks.bookmark")
+local render_bookmark = require("bookmarks.render.bookmark")
+
+local INTENT = "    "
+local M = {}
+
+local context_impl = {}
+
+---@class Bookmarks.LineContext
+---@field deep number
+---@field id number
+---@field root_name string
+
+---@class Bookmarks.TreeContext
+---@field line_contexts Bookmarks.LineContext[]
+
+
+---@param node Bookmarks.BookmarkList | Bookmarks.Bookmark
+---@param deep number
+---@return string
+function M.render_context(node, deep)
+  local node_type = domain.bookmark.get_value_type(node)
+  local icon = node.collapse and "▸" or "▾"
+  local book_icon = ""
+
+  if node_type == domain.type.BOOKMARK then
+    return string.rep(INTENT, deep) .. book_icon .. render_bookmark.render_bookmark(node)
+  else
+    return string.rep(INTENT, deep) .. icon .. node.name
+  end
+
+  return ""
+end
+
+---@param node Bookmarks.Bookmark | Bookmarks.BookmarkList
+---@param lines string[]
+---@param line_contexts Bookmarks.LineContext[]
+---@param deep number
+function M.render_tree_recursive(node, lines, line_contexts, deep, root_id)
+  local ctx = M.to_line_context(node.id, deep, root_id)
+  local line = M.render_context(node, deep)
+  table.insert(lines, line)
+  table.insert(line_contexts, ctx)
+
+  if node.collapse then
+    return
+  end
+
+  if domain.bookmark.get_value_type(node) == domain.type.BOOKMARK then
+    return
+  end
+
+  for _, child in ipairs(node.bookmarks) do
+    M.render_tree_recursive(child, lines, line_contexts, deep + 1, root_id)
+  end
+end
+
+---@param bookmark_lists Bookmarks.BookmarkList[]
+---@return Bookmarks.TreeContext, string[]
+function M.from_bookmark_lists(bookmark_lists)
+  local lines = {}
+  local line_contexts = {}
+
+  table.sort(bookmark_lists, function(a, b)
+    return a.name < b.name
+  end)
+
+  for _, bookmark_list in ipairs(bookmark_lists) do
+    M.render_tree_recursive(bookmark_list, lines, line_contexts, 0, bookmark_list.name)
+  end
+
+  local ctx = {
+    line_contexts = line_contexts,
+  }
+
+  return ctx, lines
+end
+
+---@param id string | number
+---@param deep number
+---@return Bookmarks.LineContext
+function M.to_line_context(id, deep, root_name)
```

Comment:
A `node.lua` module within the tree folder
then put 
`render_context`,
`to_line_context`
inside node.lua module

Reply:
but it's also ok to keep it this way

Reply:
do you mean rename the name of context? I just don't get it.